src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -11,80 +11,507 @@
-    ucc_tl_ucp_allreduce_sw_buf_t *getbuf = NULL;
-    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe = NULL;
-
-    // suppress "function unused" Werrors
-    ucc_tl_ucp_allreduce_sliding_window_barrier(coll_task);
-    ucc_tl_ucp_allreduce_sliding_window_allgather_free_rkeys(coll_task);
-    ucc_tl_ucp_allreduce_sliding_window_allgather_info_test(coll_task);
-    ucc_tl_ucp_allreduce_sliding_window_req_test(request, task);
-    ucc_tl_ucp_allreduce_sliding_window_test_reduction(task);
-    ucc_tl_ucp_allreduce_sliding_window_reduction(coll_task, accbuf, getbuf);
-    ucc_tl_ucp_allreduce_sliding_window_finalize(coll_task);
-    ucc_tl_ucp_allreduce_sliding_window_start(coll_task);
-    ucc_tl_ucp_allreduce_sliding_window_reset_pipeline(pipe, 0, 0);
-    ucc_tl_ucp_allreduce_sliding_window_reset_buf(accbuf);
+    ucc_tl_ucp_task_t *task    = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_rank_t         size    = (ucc_rank_t)task->subset.map.ep_num;
+    ucc_datatype_t     dtype   = TASK_ARGS(task).dst.info.datatype;
+    size_t             dt_size = ucc_dt_size(dtype);
+    uint32_t           host_team_size = size;
+    ucc_base_team_t   *base_team      = coll_task->team;
+    ucc_tl_ucp_team_t *tl_team = ucc_derived_of(base_team, ucc_tl_ucp_team_t);
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe =
+        task->allreduce_sliding_window.pipe;
+    ucc_tl_ucp_context_t          *tl_ctx    = UCC_TL_UCP_TEAM_CTX(tl_team);
+    ucc_tl_ucp_allreduce_sw_buf_t *accbuf    = &pipe->accbuf;
+    ucp_request_param_t            req_param = {0};
+    int                            i         = 0;
+    ucc_coll_task_t *allgather_task  =
+        task->allreduce_sliding_window.allgather_task;
+    ucc_coll_task_t *barrier_task    =
+        task->allreduce_sliding_window.barrier_task;
+    int              put_window_size =
+        UCC_TL_UCP_TEAM_LIB(tl_team)->
+            cfg.allreduce_sliding_window_put_window_size;
+    size_t           remaining_elems;
+    size_t           get_idx;
+    size_t           count;
+    size_t           get_offset;
+    size_t           data_size;
+    ucc_rank_t       src_rank;
+    ucc_rank_t       dst_rank;
+    void            *src_addr;
+    void            *dst_addr;
+    ucs_status_ptr_t request;
+    size_t           red_idx;
+    size_t           put_offset;
+    int              window;
+    int              put_idx;
+    ucp_ep_h         ep;
+    ucc_tl_ucp_allreduce_sw_buf_t *redbuf;
+    ucc_tl_ucp_allreduce_sw_buf_t *getbuf;
+
+    ucc_assert(host_team_size > 0);
+
+    if (barrier_task != NULL) {
+        // mark sliding window task complete once barrier finishes
+        if (barrier_task->super.status == UCC_OK) {
+            ucc_tl_ucp_put_task(
+                ucc_derived_of(task->allreduce_sliding_window.barrier_task,
+                               ucc_tl_ucp_task_t));
+            task->allreduce_sliding_window.barrier_task = NULL;
+            task->super.status                          = UCC_OK;
```

Comment:
IIUC the barrier only occurs at the end of the collective. If this is the case, this can be scheduled during initialization of allreduce (we have the ability to subscribe a task to the completion of another). See for example https://github.com/openucx/ucc/blob/b272589c0a462c3150cf7f6afe69d7d2212e4c30/src/components/tl/mlx5/alltoall/alltoall_coll.c#L876
Does it make sense ?

Reply:
Sergey had the same comment--when I talked with him about it he agreed to accept how it is and we can make a follow-up PR later so that we can make the July release. Is that ok with you as well?

Reply:
ok, sounds good!